File: NOIP/Repecharge/NOIP2010/提高组参考程序/C++/tortoise.cpp
#include <iostream>
#include <fstream>
using namespace std;

const int maxcount=40;
const int maxn=350;

int b[5],f[maxcount+2][maxcount+2][maxcount+2][maxcount+2],score[maxn+1];

int main()
{
	ifstream infile("tortoise.in",ios::in);
	ofstream outfile("tortoise.out",ios::out);
	memset(b,0,sizeof(b));
	memset(f,0,sizeof(f));
	int n,m,x,step;
	infile>>n>>m;
	for(int i=1;i<=n;i++)
		infile>>score[i];
	for(int i=1;i<=m;i++){
		infile>>x;
		b[x]++;
	}
	for(int i=1;i<=b[1]+1;i++)
		for(int j=1;j<=b[2]+1;j++)
			for(int k=1;k<=b[3]+1;k++)
				for(int l=1;l<=b[4]+1;l++){
					step=i+j*2+k*3+l*4-10+1;
					if(f[i][j][k][l]<f[i-1][j][k][l]) 
						f[i][j][k][l]=f[i-1][j][k][l];
					if(f[i][j][k][l]<f[i][j-1][k][l]) 
						f[i][j][k][l]=f[i][j-1][k][l];
					if(f[i][j][k][l]<f[i][j][k-1][l]) 
						f[i][j][k][l]=f[i][j][k-1][l];
					if(f[i][j][k][l]<f[i][j][k][l-1]) 
						f[i][j][k][l]=f[i][j][k][l-1];
					f[i][j][k][l]+=score[step];
				}
	outfile<<f[b[1]+1][b[2]+1][b[3]+1][b[4]+1]<<endl;
	infile.close();
	outfile.close();
	return 0;
}

File: NOIP/Final/NOIP2010/SeniorRefCode/Pascal/tortoise.pas
program tortoise(input,output);
  const
    maxcount=40;
    maxn=350;
  var
    f:array[-1..maxcount,-1..maxcount,-1..maxcount,-1..maxcount] of longint;
    score:array[1..maxn] of longint;
    a:array[1..4] of longint;
    n,m,i,j,k,l,step,x:longint;
  begin
    assign(input,'tortoise.in');reset(input);
    assign(output,'tortoise.out');rewrite(output);
    readln(n,m);
    for i:=1 to n do
      read(score[i]);
    readln;
    fillchar(a,sizeof(a),0);
    for i:=1 to m do begin
      read(x);
      inc(a[x]);
    end;
    fillchar(f,sizeof(f),0);
    for i:=0 to a[1] do
      for j:=0 to a[2] do
        for k:=0 to a[3] do
          for l:=0 to a[4] do begin
            step:=i+j*2+k*3+l*4+1;
            if f[i,j,k,l]<f[i-1,j,k,l] then
              f[i,j,k,l]:=f[i-1,j,k,l];
            if f[i,j,k,l]<f[i,j-1,k,l] then
              f[i,j,k,l]:=f[i,j-1,k,l];
            if f[i,j,k,l]<f[i,j,k-1,l] then
              f[i,j,k,l]:=f[i,j,k-1,l];
            if f[i,j,k,l]<f[i,j,k,l-1] then
              f[i,j,k,l]:=f[i,j,k,l-1];
            inc(f[i,j,k,l],score[step]);
          end;
    writeln(f[a[1],a[2],a[3],a[4]]);
    close(input);close(output);
  end.